### PR TITLE
fix(buffer): fix attach events

### PR DESCRIPTION
**Pull request by chemzqm**
* Not use internal isAttached
* Add buffer.isAttached method
* Unbind listener if attach failed

### CODE REVIEW

**Comment by justinmk** [comment]
ping @neovim/nodejs

File: src/api/Buffer.ts [hunk 1]
```diff
@@ -33,8 +36,17 @@ export class Buffer extends BaseApi {
...
    *        a `nvim_buf_changedtick_event`
    */
-  [ATTACH] = (sendBuffer: boolean = false, options: {} = {}) =>
-    this.request(`${this.prefix}attach`, [this, sendBuffer, options]);
+  [ATTACH] = async (
+    sendBuffer: boolean = false,
+    options: {} = {}
+  ): Promise<boolean> => {
+    if (this.client.isAttached(this)) return true;
+    return this.request(`${this.prefix}attach`, [
```
Comment (billyvg) [inline src/api/Buffer.ts]: Let's use `await` here:

```javascript
try {
 return await this.request(`${this.prefix}attach`, [this,
      sendBuffer,
      options,
    ]);
} catch(err) {
 return false;
}
```

File: src/api/Buffer.ts [hunk 2]
```diff
@@ -243,20 +255,22 @@ export class Buffer extends BaseApi {
    */
   listen(eventName: string, cb: Function): Function {
     if (!this.isAttached) {
-      this[ATTACH]();
-      this.isAttached = true;
+      this[ATTACH]().then(attached => {
```
Comment (billyvg) [inline src/api/Buffer.ts]: `await` instead of `then` here too (will also have to make this method `async`)

Comment (chemzqm) [reply src/api/Buffer.ts]: I'm fine with `async`, but that would be break change.

Comment (billyvg) [reply src/api/Buffer.ts]: Ah you're right, I didn't read down far enough, let's revert this change. Sorry about that.

**Pull request closed without merging**